### ISSUE TITLE
Incorrect semantic error reported by clang++ when using "uint16_t" together with global namespace.

### ISSUE THREAD

**Issue by lichenxuan87** [llvm/llvm-project#121457]
### Source code and error message
Error message was reported by clang++ when compiling below source code:
```
#include <stdint.h>

class CLangTester {
 public:
    uint8_t test();
};

uint8_t ::CLangTester::test() {
    return 23;
}

int main() {
    CLangTester tester;
    tester.test();
    return 0;
}

```

Error message:
```
/usr/bin/clang++ -c -I/usr/local/include -I. -Wall -g -std=c++14 namespace.cpp -o build/namespace.o
namespace.cpp:21:1: error: 'uint8_t' (aka 'unsigned char') is not a class, namespace, or enumeration
uint8_t ::CLangTester::test() {
^
1 error generated.
make: *** [Makefile:28: build/namespace.o] Error 1
```

### Version info
clang --version output:
```
clang version 10.0.0-4ubuntu1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

```

### Additional info
The error disappears when I changed 
```
uint8_t ::CLangTester::test() {
```
to 
```
uint8_t CLangTester::test() {
```

Apparently the clang++ ignored the space between `uint8_t` and `::`

**Comment by safocl** [comment]
[dcl.fct.def p1](https://eel.is/c++draft/dcl.fct.def#nt:function-definition)

> Function definitions have the form
> function-definition:
> attribute-specifier-seq opt decl-specifier-seq opt [declarator](https://eel.is/c++draft/dcl.decl.general#nt:declarator) virt-specifier-seq opt function-body
> attribute-specifier-seq opt decl-specifier-seq opt [declarator](https://eel.is/c++draft/dcl.decl.general#nt:declarator) requires-clause function-body

[dcl.decl.general p5](https://eel.is/c++draft/dcl.decl.general#nt:declarator)
> declarator:
> [ptr-declarator](https://eel.is/c++draft/dcl.decl.general#nt:ptr-declarator)
> [...]

> [ptr-declarator:](https://eel.is/c++draft/dcl.decl.general#nt:ptr-declarator)
> [noptr-declarator](https://eel.is/c++draft/dcl.decl.general#nt:noptr-declarator)
> [...]

> noptr-declarator:
> [declarator-id](https://eel.is/c++draft/dcl.decl.general#nt:declarator-id) [attribute-specifier-seq](https://eel.is/c++draft/dcl.attr.grammar#nt:attribute-specifier-seq)op
> [noptr-declarator](https://eel.is/c++draft/dcl.decl.general#nt:noptr-declarator) [parameters-and-qualifiers](https://eel.is/c++draft/dcl.decl.general#nt:parameters-and-qualifiers)

> declarator-id:
> ...opt [id-expression](https://eel.is/c++draft/expr.prim.id.general#nt:id-expression)

[expr.prim.id.general](https://eel.is/c++draft/expr.prim.id.general#nt:id-expression)
> id-expression:
> [unqualified-id](https://eel.is/c++draft/expr.prim.id.unqual#nt:unqualified-id)
> [qualified-id](https://eel.is/c++draft/expr.prim.id.qual#nt:qualified-id)

[expr.prim.id.qual](https://eel.is/c++draft/expr.prim.id.qual#nt:qualified-id)
> qualified-id:
> [nested-name-specifier](https://eel.is/c++draft/expr.prim.id.qual#nt:nested-name-specifier) template opt [unqualified-id](https://eel.is/c++draft/expr.prim.id.unqual#nt:unqualified-id)

> nested-name-specifier:
> ::
> [...]

that is, the name of a class method in the definition may contain the "::" sign in front.

**Comment by safocl** [comment]
How should this be parsed?
```cpp
struct ItIsType{};

struct Typer {
    using ItIsType = ::ItIsType;
};

class CLangTester {
 public:
    ::Typer::ItIsType test();
};

::Typer::ItIsType ::CLangTester::test() {
    return {};
}
```
https://godbolt.org/z/5TK7f6va8
:)

**Comment by zygoloid** [comment]
The relevant rule is [[basic.loolup.qual.general]/1](https://eel.is/c++draft/basic.lookup.qual#general-1.sentence-2):

> If a name, template-id, or computed-type-specifier is followed by a ::, it shall designate a namespace, class, enumeration, or dependent type, and the :: is never interpreted as a complete nested-name-specifier.

So clang is correct to reject this -- though the diagnostic could be better, especially when there's a space before the `::`.

**Comment by safocl** [comment]
> The relevant rule is [[basic.loolup.qual.general]/1](https://eel.is/c++draft/basic.lookup.qual#general-1.sentence-2):
> 
> > If a name, template-id, or computed-type-specifier is followed by a ::, it shall designate a namespace, class, enumeration, or dependent type, and the :: is never interpreted as a complete nested-name-specifier.
> 
> So clang is correct to reject this -- though the diagnostic could be better, especially when there's a space before the `::`.

this code compiles:
```cpp
class CLangTester {
 public:
    int test();
};

int ::CLangTester::test() {
    return {};
}
```
https://godbolt.org/z/TK57dYYT9

**Comment by zygoloid** [comment]
Yes. `int` is a keyword, not a "name, template-id, or computed-type-specifier". But an identifier such as `uint16_t` followed by a `::` is always treated as forming a qualified access into `uint16_t`.

**Comment by shafik** [comment]
This [cwg1828](https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1828)

**Comment by lichenxuan87** [comment]
> > The relevant rule is [[basic.loolup.qual.general]/1](https://eel.is/c++draft/basic.lookup.qual#general-1.sentence-2):
> > > If a name, template-id, or computed-type-specifier is followed by a ::, it shall designate a namespace, class, enumeration, or dependent type, and the :: is never interpreted as a complete nested-name-specifier.
> > 
> > 
> > So clang is correct to reject this -- though the diagnostic could be better, especially when there's a space before the `::`.
> 
> this code compiles:
> 
> class CLangTester {
>  public:
>     int test();
> };
> 
> int ::CLangTester::test() {
>     return {};
> }
> https://godbolt.org/z/TK57dYYT9

There's a space between "uint16_t" and "::". I don't think the rule applies, as space is a common semantic separator.

**Comment by safocl** [comment]
> The relevant rule is [[basic.loolup.qual.general]/1](https://eel.is/c++draft/basic.lookup.qual#general-1.sentence-2):
> 
> > If a name, template-id, or computed-type-specifier is followed by a ::, it shall designate a namespace, class, enumeration, or dependent type, and the :: is never interpreted as a complete nested-name-specifier.
> 
> So clang is correct to reject this -- though the diagnostic could be better, especially when there's a space before the `::`.

[[basic.def]/1](https://eel.is/c++draft/basic.lookup#general-1.sentence-1)
> [Name lookup](https://eel.is/c++draft/basic.lookup#def:lookup,name) associates the use of a name with a set of declarations ([[basic.def]](https://eel.is/c++draft/basic.def)) of that name[.](https://eel.is/c++draft/basic.lookup#general-1.sentence-1)

This means that the rules of "name lookup" do not apply in this case (for definitions).

**Comment by safocl** [comment]
> [[basic.def]/1](https://eel.is/c++draft/basic.lookup#general-1.sentence-1)
> 
> > [Name lookup](https://eel.is/c++draft/basic.lookup#def:lookup,name) associates the use of a name with a set of declarations ([[basic.def]](https://eel.is/c++draft/basic.def)) of that name[.](https://eel.is/c++draft/basic.lookup#general-1.sentence-1)

upd. Or it looks like I misunderstood this phrase

**Comment by killerbee13** [comment]
> There's a space between "uint16_t" and "::". I don't think the rule applies, as space is a common semantic separator.

The space in this context is not semantically significant, although it may be useful for tailoring the diagnostic. In C++, the only semantic meaning of whitespace is to separate tokens, so `a::b` and `a :: b` are identical in meaning.

**Comment by shafik** [comment]
> > There's a space between "uint16_t" and "::". I don't think the rule applies, as space is a common semantic separator.
> 
> The space in this context is not semantically significant, although it may be useful for tailoring the diagnostic. In C++, the only semantic meaning of whitespace is to separate tokens, so `a::b` and `a :: b` are identical in meaning.

To be more specific after [lex phase 7](https://eel.is/c++draft/lex.phases#1.7):

> ... Whitespace characters separating tokens are no longer significant. ...
